database.rules.json argument path bug

<!-- DO NOT DELETE
validate_template=true
template_path=.github/ISSUE_TEMPLATE/bug_report.md
-->

### Environment info

<!-- What version of the Firebase CLI (`firebase-tools`) are you using? Note that your issue may already be fixed in the latest versions. The latest version can be found at https://github.com/firebase/firebase-tools/releases -->

<!-- Output of `firebase --version` -->

**firebase-tools:** 8.7.0

<!-- e.g. macOS, Windows, Ubuntu -->

**Platform:** macOS Catalina 10.15.5

### Test case

<!-- Provide a minimal, complete, and verifiable example (http://stackoverflow.com/help/mcve) -->
**npm run serve function** will not work if you remove **--only functions** argument and add **npm i firebase-tools** to your project. It will complain about **database.rules.json** path. Log from **firebase-debug.log** last line:
```
Error: ENOENT: no such file or directory, open 'database.rules.json'
    at Object.openSync (fs.js:458:3)
    at Object.readFileSync (fs.js:360:35)
    at DatabaseEmulator.<anonymous> (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:69:59)
    at Generator.next (<anonymous>)
    at /Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:4:12)
    at DatabaseEmulator.connect (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:63:16)
    at Object.<anonymous> (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/controller.js:421:32)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/controller.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[error] 
[error] Error: An unexpected error has occurred.
```
If you will follow up logic in the **controller**, it seems like **firebase-tools** are sending args and realtime database block is rewriting rules fields in them. Which leads to error in **databaseEmulator.js** on line:
` yield this.updateRules(c.instance, fs.readFileSync(c.rules, "utf8").toString());`
In this case, **c.rules** is **'database.rules.json'** string, which is of course **not the right relative path**.
Interesting enough is that this error will only appear if you **use script npm run serve** (corrected to npm run build && firebase emulators:start) , but if you will run **npm run build && firebase emulators:start** in **terminal** directly it will work, because it's not sending any args as project id and so on.
This args rules console logged:
` rules: [ { rules: 'database.rules.json', instance: *ProjectID* } ]`

### Steps to reproduce

<!-- Provide the steps needed to reproduce the issue with the above test case. -->
```
mkdir anyName
cd anyName
firebase init
select everything without firebase hosting and select typescript
change npm run serve command from 
 "serve": "npm run build && firebase emulators:start --only functions"
to
 "serve": "npm run build && firebase emulators:start"
npm i firebase-tools
run npm run serve command and receive an error
run in terminal npm run build && firebase emulators:start and there will be no error
```

### Expected behavior

<!-- What is the expected behavior? -->
I expect emulators to work with the firebase-tools package, the question is if it's intended or not? Currently, I wanted to use this package in scheduled function to delete recursively storage folders, as it's advised here https://firebase.google.com/docs/firestore/solutions/delete-collections, of course, there are other possibilities to do that, but I'm just trying to be sure about this particular problem.

### Actual behavior
Emulator UI has exited upon receiving signal: SIGINT
Error: An unexpected error has occurred.

<!-- Run the command with --debug flag, and include the logs below. -->
```
✔  functions[helloWorld]: http function initialized (http://localhost:5001/projectId/us-central1/helloWorld). {"metadata":{"emulator":{"name":"functions"},"message":"\u001b[1mhttp\u001b[22m function initialized (http://localhost:5001/projectId/us-central1/helloWorld)."}}
i  emulators: Shutting down emulators. {"metadata":{"emulator":{"name":"hub"},"message":"Shutting down emulators."}}
i  hub: Stopping emulator hub {"metadata":{"emulator":{"name":"hub"},"message":"Stopping emulator hub"}}
i  ui: Stopping Emulator UI {"metadata":{"emulator":{"name":"ui"},"message":"Stopping Emulator UI"}}
i  logging: Stopping Logging Emulator {"metadata":{"emulator":{"name":"logging"},"message":"Stopping Logging Emulator"}}
i  functions: Stopping Functions Emulator {"metadata":{"emulator":{"name":"functions"},"message":"Stopping Functions Emulator"}}
⚠  Your function was killed because it raised an unhandled error. {"metadata":{"emulator":{"name":"functions"},"message":"Your function was killed because it raised an unhandled error."}}
i  firestore: Stopping Firestore Emulator {"metadata":{"emulator":{"name":"firestore"},"message":"Stopping Firestore Emulator"}}
i  database: Stopping Database Emulator {"metadata":{"emulator":{"name":"database"},"message":"Stopping Database Emulator"}}
i  pubsub: Stopping Pub/Sub Emulator {"metadata":{"emulator":{"name":"pubsub"},"message":"Stopping Pub/Sub Emulator"}}
[2020-08-11T18:31:01.228Z] [worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: IDLE {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: IDLE"}}
[2020-08-11T18:31:01.229Z] *** shutting down gRPC server since JVM is shutting down
 {"metadata":{"emulator":{"name":"firestore"},"message":"*** shutting down gRPC server since JVM is shutting down\n"}}
⚠  Emulator UI has exited upon receiving signal: SIGINT 
[2020-08-11T18:31:01.229Z] *** shutting down gRPC server since JVM is shutting down
 {"metadata":{"emulator":{"name":"pubsub"},"message":"*** shutting down gRPC server since JVM is shutting down\n"}}
[2020-08-11T18:31:01.229Z] 20:31:01.228 [Thread-1] INFO com.firebase.server.forge.App$ - Attempting graceful shutdown.
 {"metadata":{"emulator":{"name":"database"},"message":"20:31:01.228 [Thread-1] INFO com.firebase.server.forge.App$ - Attempting graceful shutdown.\n"}}
[2020-08-11T18:31:01.230Z] [worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: exited {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: exited"}}
[2020-08-11T18:31:01.230Z] [worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: FINISHED {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-ca20a50e-2437-4b6c-9b33-932c1f474452]: FINISHED"}}
[2020-08-11T18:31:01.232Z] *** server shut down
 {"metadata":{"emulator":{"name":"firestore"},"message":"*** server shut down\n"}}
[2020-08-11T18:31:01.233Z] *** server shut down
 {"metadata":{"emulator":{"name":"pubsub"},"message":"*** server shut down\n"}}
[2020-08-11T18:31:01.234Z] 20:31:01.234 [Thread-1] INFO com.firebase.server.forge.App$ - Graceful shutdown complete.
 {"metadata":{"emulator":{"name":"database"},"message":"20:31:01.234 [Thread-1] INFO com.firebase.server.forge.App$ - Graceful shutdown complete.\n"}}
[2020-08-11T18:31:01.603Z] Error: ENOENT: no such file or directory, open 'database.rules.json'
    at Object.openSync (fs.js:458:3)
    at Object.readFileSync (fs.js:360:35)
    at DatabaseEmulator.<anonymous> (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:69:59)
    at Generator.next (<anonymous>)
    at /Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:4:12)
    at DatabaseEmulator.connect (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/databaseEmulator.js:63:16)
    at Object.<anonymous> (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/controller.js:421:32)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/*****/Desktop/bug-test/functions/node_modules/firebase-tools/lib/emulator/controller.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)

Error: An unexpected error has occurred.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! functions@ serve: `npm run build && firebase emulators:start --debug`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the functions@ serve script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/*****/.npm/_logs/2020-08-11T18_31_01_890Z-debug.log
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,4 @@
 - Fixes an issue where immutable params were not sent during ext:configure.
 - Includes latest features and improvements from production in the Firestore Emulator.
 - Fixes issue where all database functions triggered on the default namespace (#2501)
+- Fixes issue where rules paths were not normalized before reading (#2544)

--- a/src/database/rulesConfig.ts
+++ b/src/database/rulesConfig.ts
@@ -1,4 +1,5 @@
 import { FirebaseError } from "../error";
+import * as Config from "../config";
 import * as logger from "../logger";
 
 export interface RulesInstanceConfig {
@@ -10,6 +11,22 @@ interface DatabaseConfig {
   rules: string;
   target?: string;
   instance?: string;
+}
+
+/**
+ * Convert the relative paths in the config into absolute paths ready to be read.
+ */
+export function normalizeRulesConfig(
+  rulesConfig: RulesInstanceConfig[],
+  options: any
+): RulesInstanceConfig[] {
+  const config = options.config as Config;
+  return rulesConfig.map((rc) => {
+    return {
+      instance: rc.instance,
+      rules: config.path(rc.rules),
+    };
+  });
 }
 
 export function getRulesConfig(projectId: string, options: any): RulesInstanceConfig[] {

--- a/src/emulator/controller.ts
+++ b/src/emulator/controller.ts
@@ -4,6 +4,7 @@ import * as fs from "fs";
 import * as path from "path";
 import * as http from "http";
 
+import * as Config from "../config";
 import * as logger from "../logger";
 import * as track from "../track";
 import * as utils from "../utils";
@@ -457,10 +458,11 @@ export async function startAll(options: any, noUi: boolean = false): Promise<voi
       args.seed_from_export = exportMetadataFilePath;
     }
 
-    const rulesLocalPath = options.config.get("firestore.rules");
+    const config = options.config as Config;
+    const rulesLocalPath = config.get("firestore.rules");
     let rulesFileFound = false;
     if (rulesLocalPath) {
-      const rules: string = path.join(options.projectRoot, rulesLocalPath);
+      const rules: string = config.path(rulesLocalPath);
       rulesFileFound = fs.existsSync(rules);
       if (rulesFileFound) {
         args.rules = rules;
@@ -502,7 +504,10 @@ export async function startAll(options: any, noUi: boolean = false): Promise<voi
       auto_download: true,
     };
 
-    const rc = dbRulesConfig.getRulesConfig(projectId, options);
+    const rc = dbRulesConfig.normalizeRulesConfig(
+      dbRulesConfig.getRulesConfig(projectId, options),
+      options
+    );
     logger.debug("database rules config: ", JSON.stringify(rc));
 
     args.rules = rc;

--- a/src/emulator/controller.ts
+++ b/src/emulator/controller.ts
@@ -520,7 +520,7 @@ export async function startAll(options: any, noUi: boolean = false): Promise<voi
       );
     } else {
       for (const c of rc) {
-        const rules: string = path.join(options.projectRoot, c.rules);
+        const rules: string = c.rules;
         if (!fs.existsSync(rules)) {
           databaseLogger.logLabeled(
             "WARN",

